### PR TITLE
#238 enforce login and tag index

### LINKED ISSUE #238: Enforce Tag Index collaboration
# Problem

We have all these features to make sure librarianettes are _doing the right thing_. For example, tag validation, and content warning reminders.

However, if a librarianette just doesn't log in to Google or set the tag index, they can freely mess around with library data with no safety nets.

# Solution

A config that enforces that people are logged in and have a Tag Index set to actually use librarything. And? Defaults to ON!

If people are not logged with a tag index set, the whole librarything website is not useable. A modal covers the screen saying "please log in and set index"

This _can_ be disabled in the options, but will give you a BIG warning if you do.

### PR DESCRIPTION
**Pull request by adirar111**
modal stays up until you're logged in and the tag index is set
resolves #238

Kapture 2023-01-06 at 23 18 03

### CODE REVIEW

File: src/sass/options.sass [hunk 1]
```diff
@@ -51,6 +52,72 @@ body
...
+
+
+/* Rounded sliders */
+.slider.round
+	border-radius: 34px
+
+
+.slider.round:before
+	border-radius: 50%
+
```
Comment (adirar111) [inline src/sass/options.sass:120]: all stolen from w3 schools

File: src/ts/common/entities/config.ts [hunk 2]
```diff
@@ -13,20 +13,23 @@ enum ConfigKey {
 	SizeData = "size-data",
 	LatestRender = "last-render",
 	FormData = "form-data",
+	EnforceAuthAndTagIndex = "enforce-auth-tag-index",
```
Comment (adirar111) [inline src/ts/common/entities/config.ts:16]: this feels like a bad name but idk what else to call it

File: src/ts/common/ui/modal.ts [hunk 3]
```diff
@@ -24,9 +24,10 @@ interface ModalInput extends ModalElement {
 interface ModalOptions {
 	text: string;
 	subText?: string[];
-	elements: (ModalButton | ModalInput)[];
+	elements?: (ModalButton | ModalInput)[];
```
Comment (adirar111) [inline src/ts/common/ui/modal.ts:27]: idk if we want to do this, i just didn't need it for my use case

File: src/ts/common/ui/modal.ts [hunk 4]
```diff
@@ -96,12 +97,19 @@ const addOnClick = (element: HTMLElement, exit: () => void, onClick?: () => Prom
...
+	onCancel,
+	colour,
+	exitable = true,
+}: ModalOptions): (() => HTMLDivElement) => {
 	const exit = () => document.body.removeChild(overlay);
 
 	const overlay = createOverlay();
 	overlay.classList.add("modal");
-	addOnClick(overlay, exit, onCancel);
+	exitable && addOnClick(overlay, exit, onCancel);
```
Comment (adirar111) [inline src/ts/common/ui/modal.ts:112]: prevent dismissing on click away unless it's exitable

File: src/ts/common/ui/modal.ts [hunk 5]
```diff
@@ -115,6 +123,8 @@ const createModal = ({text, subText, elements, onCancel, colour}: ModalOptions):
 	modal.append(textContainer, elementContainer);
 	overlay.append(modal);
 	document.body.appendChild(overlay);
+
+	return exit;
```
Comment (adirar111) [inline src/ts/common/ui/modal.ts:127]: i needed this to exit when certain conditions are met, so i'm returning it here

File: src/ts/content/extensions/enforceAuthAndTagIndex.ts [hunk 6]
```diff
@@ -0,0 +1,37 @@
...
+				exitable: false,
+			});
+
+			onBackgroundEvent(BackgroundEvent.CompletedAuth, async () => {
+				if (await isSpreadSheetLinkSet()) {
+					exit();
+				}
+			});
+
+			onceBackgroundEvent(BackgroundEvent.AddedSheetLink, async () => {
```
Comment (adirar111) [inline src/ts/content/extensions/enforceAuthAndTagIndex.ts:30]: still not super sure why we want `onceBackgroundEvent` for the tag index here and `onBackgroundEvent` for auth

File: src/sass/options.sass (hunk 1 above)
Comment (braxtonhall) [reply src/sass/options.sass:120]: _nice_

File: src/html/options.html [hunk 7]
```diff
@@ -23,6 +23,16 @@ <h4 class="label"><a href="https://betterlibrarything.com/docs/librarian/tag-ind
 	<button id="save-tag-index" class="log-button colour-default" disabled>Save</button>
 </div>
 
+<div class="form">
+	<h4 class="label">Enfore logging in and setting tag index before using LibraryThing?</h4>
```
Comment (braxtonhall) [inline src/html/options.html:27]: Can we use name it `Tag Index`? Just to be consistent with the rest of our docs

File: src/ts/common/entities/config.ts (hunk 2 above)
Comment (braxtonhall) [reply src/ts/common/entities/config.ts:16]: This is fine but we could do `EnforceTagIndexAccess`? Dunno if that's better

File: src/ts/common/ui/modal.ts (hunk 3 above)
Comment (braxtonhall) [reply src/ts/common/ui/modal.ts:27]: Actually, I think a button to open the options would be a good idea! You can use `invokeWorker(OpenOptions)` or something like that to open them.

File: src/ts/content/extensions/enforceAuthAndTagIndex.ts [hunk 8]
```diff
@@ -0,0 +1,37 @@
...
+import {BackgroundEvent} from "../../common/backgroundEvent";
+import {isAuthorized} from "./author/util/isAuthorized";
+import {createModal} from "../../common/ui/modal";
+import {UIColour} from "../../common/ui/colour";
+
+const isSpreadSheetLinkSet = async () => config.get(ConfigKey.SpreadsheetLink);
+
+window.addEventListener("pageshow", async () => {
+	if (await config.get(ConfigKey.EnforceAuthAndTagIndex)) {
+		if (!(await isAuthorized()) || !(await isSpreadSheetLinkSet())) {
```
Comment (braxtonhall) [inline src/ts/content/extensions/enforceAuthAndTagIndex.ts:13]: I _think_  you're recreating `onLogged`.
`onLoggedIn` is called when both of these conditions are met (like when you invoke exit)
and `onLoggedOut` is called if one of the conditions aren't met (when you invoke exit)

If you use `onLogged`, you also get the advantage that the modal would pop up again if you log out in another tab

Comment (braxtonhall) [reply src/ts/content/extensions/enforceAuthAndTagIndex.ts:13]: Just a thought, but not necessary -- if we're going to make things react to actions in other places like with onLogged, you _could_ even add `onBackgroundEvent(EnableEnforceAccess)` or something, that re-triggers this pageshow callback

```typescript
const callbackWithABetterName = () => {
	// ... ur code
};

onBackgroundEvent(BackgroundEvent.Access, callbackWithABetterName);
window.addEventListener("pageshow", callbackWithABetterName);
```

File: src/ts/content/extensions/enforceAuthAndTagIndex.ts (hunk 6 above)
Comment (braxtonhall) [reply src/ts/content/extensions/enforceAuthAndTagIndex.ts:30]: yeah i don't think this implementation needs `once` specifically... I had something specific in my brain at the time I was talking about it involving loaders etc that would have been TERRIBLE for the user

File: src/html/options.html (hunk 7 above)
Comment (braxtonhall) [reply src/html/options.html:27]: Actually this could say `Enforce Tag Index Access` too.

Comment (braxtonhall) [reply src/html/options.html:27]: (also typo: "Enfore")

File: src/ts/content/extensions/enforceAuthAndTagIndex.ts (hunk 8 above)
Comment (braxtonhall) [reply src/ts/content/extensions/enforceAuthAndTagIndex.ts:13]: hmmm but the downside is that `onLoggedOut` will get called multiple times........ which would mean multiple overlays. Let's leave it and think about it.

Might want to consider logic deduplicating in the future

**Review by braxtonhall — CHANGES_REQUESTED** [review]
Options UI looks _peeeerfect_. bless w3 schools lol
kiiiinda missss the left-justified text though. probably would look clunky with all the text

I do think the Modal needs a button though. Not everyone will know how to open the extension options. The buttons all exit the modal. if I were u I think the easiest thing to do would be to `onClick` of the button also reopen the same modal with ✨ _recursion_ ✨

File: src/ts/common/ui/modal.ts (hunk 3 above)
Comment (braxtonhall) [reply src/ts/common/ui/modal.ts:27]: Gonna use this in the docs
image

Comment (braxtonhall) [reply src/ts/common/ui/modal.ts:27]: aaa! `tag index` to `Tag Index` here too

Comment (braxtonhall) [reply src/ts/common/ui/modal.ts:27]: ok gonna use _this_ in the documentation

image

File: src/html/options.html (hunk 7 above)
Comment (braxtonhall) [reply src/html/options.html:27]: LAST REQUEST: can we spawn in a big ugly warning in this form when you disable this feature?

Like... just an element that defaults to display:none, and is shown iff toggle is set to false?

**Comment by adirar111** [comment]
updated flow

(the button is purple, i just don't feel like rerecording)

Kapture 2023-01-07 at 02 24 04

File: src/html/options.html [hunk 9]
```diff
@@ -24,7 +24,8 @@ <h4 class="label"><a href="https://betterlibrarything.com/docs/librarian/tag-ind
 </div>
 
 <div class="form">
-	<h4 class="label">Enfore logging in and setting tag index before using LibraryThing?</h4>
+	<h4 class="label">Enforce Tag Index Access</h4>
+	<h4 id="enforce-toggle-warning" class="warning">WARNING: This config option is HIGHLY reccomended to avoid any unwanted and unintended changes to the catalog</h4>
```
Comment (adirar111) [inline src/html/options.html:28]: i can't spell

File: src/ts/content/extensions/enforceTagIndexAccess.ts [hunk 10]
```diff
@@ -0,0 +1,52 @@
...
+		};
+
+		onLogged({
+			onLogIn: () => {
+				removeOverlay();
+			},
+			onLogOut: async () => {
+				if (!isOverlayRendered()) {
+					createWarningModal();
+				}
```
Comment (adirar111) [inline src/ts/content/extensions/enforceTagIndexAccess.ts:45]: to avoid multiple of these modals/overlays because of the recursion

**Review by braxtonhall — APPROVED** [review]
aight squash it

**Comment by braxtonhall** [comment]
ngl i might give it a pass in 2.0.1, but i want this thing available to natalie as asap as possible

**Pull request merged**